draft-ietf-httpbis-message-signatures.md

```diff
@@ -426,58 +484,63 @@ Application-specific requirements are expected and encouraged.  When an applicat
+## The 'Signature' HTTP Header
+The `Signature` HTTP header field is a Dictionary Structured Header {{StructuredFields}} containing zero or more message signatures generated from content within the HTTP message. Each member's name is a signature identifier that is present as a member name in the `Signature-Input` Structured Header within the HTTP message. Each member's value is a Byte Sequence containing the signature value for the message signature identified by the member name. Any member in the `Signature` HTTP header field that does not have a corresponding member in the HTTP message's `Signature-Input` HTTP header field MUST be ignored.
 
-## Example
+## Examples
 
-The following is a non-normative example Signature header field representing the signature in {{example-sig-value}}:
+The following is a non-normative example of `Signature-Input` and `Signature` HTTP header fields representing the signature in {{example-sig-value}}:
 
 ~~~
-Signature: keyId="test-key-b", algorithm="rsa-sha256",
-    created=1402170695, expires=1402170995,
-    headers="(request-target) (created) host date cache-control
-        x-emptyheader x-example",
-    signature="T1l3tWH2cSP31nfuvc3nVaHQ6IAu9YLEXg2pCeEOJETXnlWbgKtBTa
-        XV6LNQWtf4O42V2DZwDZbmVZ8xW3TFW80RrfrY0+fyjD4OLN7/zV6L6d2v7uB
-        puWZ8QzKuHYFaRNVXgFBXN3VJnsIOUjv20pqZMKO3phLCKX2/zQzJLCBQvF/5
-        UKtnJiMp1ACNhG8LF0Q0FPWfe86YZBBxqrQr5WfjMu0LOO52ZAxi9KTWSlceJ
-        2U361gDb7S5Deub8MaDrjUEpluphQeo8xyvHBoNXsqeax/WaHyRYOgaW6krxE
-        GVaBQAfA2czYZhEA05Tb38ahq/gwDQ1bagd9rGnCHtAg=="
+Signature-Input: sig1=(*request-target, *created, host, date,
+    cache-control, x-empty-header, x-example); keyId="test-key-a";
+    alg=hs2019; created=1402170695; expires=1402170995
+Signature: sig1=:K2qGT5srn2OGbOIDzQ6kYT+ruaycnDAAUpKv+ePFfD0RAxn/1BUe
+    Zx/Kdrq32DrfakQ6bPsvB9aqZqognNT6be4olHROIkeV879RrsrObury8L9SCEibe
+    oHyqU/yCjphSmEdd7WD+zrchK57quskKwRefy2iEC5S2uAH0EPyOZKWlvbKmKu5q4
+    CaB8X/I5/+HLZLGvDiezqi6/7p2Gngf5hwZ0lSdy39vyNMaaAT0tKo6nuVw0S1MVg
+    1Q7MpWYZs0soHjttq0uLIA3DIbQfLiIvK6/l0BdWTU7+2uQj7lBkQAsFZHoA96ZZg
+    FquQrXRlmYOh+Hx5D9fJkXcXe5tmAg==:
+~~~
+
+Since `Signature-Input` and `Signature` are both defined as Dictionary Structured Headers, they can be used to easily include multiple signatures within the same HTTP message. For example, a signer may include multiple signatures signing the same content with different keys and/or algorithms to support verifiers with different capabilities, or a reverse proxy may include information about the client in header fields when forwarding the request to a service host, and may also include a signature over those fields and the client's signature. The following is a non-normative example of header fields a reverse proxy might add to a forwarded request that contains the signature in the above example:
+
+~~~
+X-Forwarded-For: 192.0.2.123
+Signature-Input: reverse_proxy_sig=(*created, host, date,
```

Comment:
I like `Signature-Input`! Do you think it would be useful to separate the signed headers and signature parameters (*alg, *created, *expires, ...)?

Reply:
That's already done through the use of the structured elements and being able to "name" the signature input fields.

Reply:
I think we can actually replace content identifiers for signature metadata with signing the whole Signature-Input member (e.g., `signature-input:reverse_proxy_sig`). What remains then is header fields and identifiers for other protocol elements like `*request-target` (and similar likely-to-be-defined identifiers like `*method`, `*path`, etc.). I think splitting those out would add complexity without any clear benefit. It would also complicate the field format, since a dictionary member can't contain two lists.